src/ast.rs

```diff
@@ -0,0 +1,669 @@
+use super::jsonschema;
+use std::collections::HashMap;
+
+#[derive(Serialize, Deserialize, Debug, Copy, Clone)]
+#[serde(rename_all = "lowercase")]
+pub enum Atom {
+    Boolean,
+    Integer,
+    Number,
+    String,
+    JSON,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Object {
+    pub fields: HashMap<String, Box<Tag>>,
```

Comment:
It doesn't look like the Box is required, but this is consistent with the other structures that need it to avoid infinite recursion when type checking.